File: csharp/pig-latin/PigLatin.cs
using System;
using System.Linq;

public static class PigLatin
{
    private static readonly string[] vowels = new [] 
        { "a", "e", "i", "o", "u", "xr", "yt" };
    private const string special = "qu";

    public static string Translate(string word)
    {
        if(word.IndexOf(' ') > -1)
            return string.Join(' ', word.Split(' ').Select(Translate));

        var startingVowel = vowels.FirstOrDefault(word.StartsWith);
        if(startingVowel != null)
            return word + "ay";

        var specialIndex = word.IndexOf(special);
        if(specialIndex != -1)
            return 
                word.Substring(specialIndex + special.Length)
                + word.Substring(0, specialIndex)
                + special + "ay";

        var firstVowel = vowels
            .Select(v => word.IndexOf(v))
            .Where(i => i != -1)
            .DefaultIfEmpty(0)
            .Min();
        var consonantSound = firstVowel == 0 
            ? word : word.Substring(0, firstVowel);
        consonantSound = consonantSound.Length > 1 
            ? consonantSound.TrimEnd('y') : consonantSound;

        return $"{word.Substring(consonantSound.Length)}{consonantSound}ay";
    }
}

File: go/pig-latin/pig_latin.go
package piglatin

import "strings"

var vowels = []string{"a", "e", "i", "o", "u", "xr", "yt"}

func Sentence(source string) string {
	if strings.Contains(source, " ") {
		words := strings.Split(source, " ")
		result := ""
		for _, w := range words {
			result += " " + Sentence(w)
		}
		return strings.Trim(result, " ")
	}

	for _, v := range vowels {
		if strings.HasPrefix(source, v) {
			return source + "ay"
		}
	}

	special := strings.Index(source, "qu")
	if special != -1 {
		return source[special+2:] + source[:special] + "quay"
	}

	firstVowel := len(source)
	for _, v := range vowels {
		i := strings.Index(source, v)
		if i != -1 && i < firstVowel {
			firstVowel = i
		}
	}
	if firstVowel == len(source) {
		firstVowel = 0
	}

	consonantSound := source
	if firstVowel != 0 {
		consonantSound = source[:firstVowel]
	}

	if len(consonantSound) > 1 {
		consonantSound = strings.TrimRight(consonantSound, "y")
	}

	return source[len(consonantSound):] + consonantSound + "ay"
}
